stm32-modules/thermocycler-refresh/firmware/system/system_hardware.c

```diff
@@ -86,3 +125,23 @@ void system_hardware_enter_bootloader(void) {
         : "r" (*sysmem_boot_loc)
         : "memory"  );
 }
+
+bool system_front_button_pressed(void) {
+    // Active low button, passively pulled high
+    return HAL_GPIO_ReadPin(FRONT_BUTTON_IN_PORT, FRONT_BUTTON_IN_PIN) 
+                == GPIO_PIN_RESET;
+}
+
+void system_front_button_callback(void) {
+    if(__HAL_GPIO_EXTI_GET_IT(FRONT_BUTTON_IN_PIN) != 0x00u) {
+        __HAL_GPIO_EXTI_CLEAR_IT(FRONT_BUTTON_IN_PIN);
+        uint32_t new_tick = HAL_GetTick();
```

Comment:
we should probably think about overflow here - it's an appliance that might be left on all the time.